http3/client.go

```diff
@@ -98,32 +110,92 @@ func (c *client) dial() error {
 	}
 
 	// run the sesssion setup using 0-RTT data
-	go func() {
-		if err := c.setupSession(); err != nil {
-			c.logger.Debugf("Setting up session failed: %s", err)
-			c.session.CloseWithError(quic.ErrorCode(errorInternalError), "")
-		}
-	}()
+	// go func() {
+	if rerr := c.setupSession(); rerr.HasError() {
+		c.logger.Debugf("Setting up session failed: %s", rerr)
+		c.session.CloseWithError(quic.ErrorCode(rerr.connErr), "")
+	}
+	// }()
 
 	return nil
 }
 
-func (c *client) setupSession() error {
+func (c *client) setupSession() requestError {
 	// open the control stream
 	str, err := c.session.OpenUniStream()
 	if err != nil {
-		return err
+		return newConnError(errorMissingSettings, err)
 	}
 	buf := &bytes.Buffer{}
 	// write the type byte
 	buf.Write([]byte{0x0})
 	// send the SETTINGS frame
 	(&settingsFrame{}).Write(buf)
 	if _, err := str.Write(buf.Bytes()); err != nil {
-		return err
+		return newConnError(errorMissingSettings, err)
 	}
 
-	return nil
+	controlStreamIn, err := c.session.AcceptUniStream(context.Background())
+	if err != nil {
+		c.logger.Debugf("Accepting the incoming control stream failed.")
+		return newConnError(errorInternalError, err)
+	}
+
+	br, ok := controlStreamIn.(byteReader)
+	if !ok {
+		br = &byteReaderImpl{controlStreamIn}
+	}
+	t, err := utils.ReadVarInt(br)
+	if t != 0x0 {
+		c.logger.Debugf("First stream must be a control stream")
```

Comment:
The first stream doesn't necessarily need to be the control stream.

Reply:
I can change it, but the only reason I made it like that was because of this line in the spec: 
"Each side MUST initiate a single control stream at the beginning of the connection".

Reply:
That's exactly what the type byte is there for. Just because you have to do it at the beginning doesn't mean that this necessarily is the first stream.